ext/curl/interface.c

```diff
@@ -757,6 +763,59 @@ static size_t curl_xferinfo(void *clientp, curl_off_t dltotal, curl_off_t dlnow,
 /* }}} */
 #endif
 
+#if LIBCURL_VERSION_NUM >= 0x075400
+static int curl_ssh_hostkeyfunction(void *clientp, int keytype, const char *key, size_t keylen)
+{
+	php_curl *ch = (php_curl *)clientp;
+	php_curl_callback *t = ch->handlers.sshhostkey;
+	int rval = CURLKHMATCH_MISMATCH; /* cancel connection in case of an exception */
+
+#if PHP_CURL_DEBUG
+	fprintf(stderr, "curl_ssh_hostkeyfunction() called\n");
+	fprintf(stderr, "clientp = %x, keytype = %d, key = %s, keylen = %zu\n", clientp, keytype, key, keylen);
+#endif
+
+	zval argv[4];
+	zval retval;
+	zend_result error;
+	zend_fcall_info fci;
+
+	GC_ADDREF(&ch->std);
+	ZVAL_OBJ(&argv[0], &ch->std);
+	ZVAL_LONG(&argv[1], keytype);
+	ZVAL_STRING(&argv[2], key);
+	ZVAL_LONG(&argv[3], keylen);
+
+	fci.size = sizeof(fci);
+	ZVAL_COPY_VALUE(&fci.function_name, &t->func_name);
+	fci.object = NULL;
+	fci.retval = &retval;
+	fci.param_count = 4;
+	fci.params = argv;
+	fci.named_params = NULL;
+
+	ch->in_callback = 1;
+	error = zend_call_function(&fci, &t->fci_cache);
+	ch->in_callback = 0;
+	if (error == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "Cannot call the CURLOPT_SSH_HOSTKEYFUNCTION");
+	} else if (!Z_ISUNDEF(retval)) {
+		_php_curl_verify_handlers(ch, /* reporterror */ true);
+		zend_long retval_long = zval_get_long(&retval);
+		if (retval_long == CURLKHMATCH_OK || retval_long == CURLKHMATCH_MISMATCH) {
+			rval = retval_long;
+		} else {
+			zend_value_error("The CURLOPT_SSH_HOSTKEYFUNCTION callback must return either CURLKHMATCH_OK or CURLKHMATCH_MISMATCH");
```

Comment:
I'm not sure a `zend_value_error` is allowed for a return value.

Reply:
Yes I was wondering that too. My gut feeling says no. kocsismate talked about the requirements of the function and that is why I added the exception. Since the function only allows two possible return values I wonder what we should do if something is returned that isn't one of the allowed values...

Reply:
I had a look at the curl code and it looks like this :

```
rc = data->set.ssh_hostkeyfunc(data->set.ssh_hostkeyfunc_userp, keytype, remotekey, keylen);
if(rc!= CURLKHMATCH_OK) {
```
We could send a E_WARNING to the user saying that the return code was not one of the two allowed and that `CURLKHMATCH_MISMATCH` will be used. What do you think ?

Reply:
Okay sounds good. Thanks for the help.
I'll do that when I have time :)

Reply:
Yes, `ValueError` always refers to arguments, not return values. A simple `TypeError` or `Error` should be thrown instead. I don't like triggering a warning, because it's a programmatic error if an invalid value is returned, and we do not have to respect BC here. So let's be strict

Reply:
OK got it! In case it's not a long, I'll use a TypeError. If it's a long but not one of the allowed values I'll use an Error (because the type is right this time).

Reply:
OK, I did some quick investigation, and I found one example where `ValueError` refers to a property (`zend_value_error("SNMP::$max_oids must be greater than 0 or null");`). But otherwise I couldn't find any example where a `ValueError` doesn't refer to arguments. However, if we think through... If `TypeError`s are used for return values, why `ValueError`s shouldn't... So I'm uncertain at this point.

Reply:
Using ValueError for a return value would be in violation of the documentation though:

> A ValueError is thrown when the type of an argument is correct but the value of it is incorrect. 

From: https://www.php.net/manual/en/class.valueerror.php

Reply:
Yeah,I saw it, but in this case, the manual should be adjusted to the code :) btw I don't mind much what exception we end up with. cc @Girgias Do you have any suggestion?

Reply:
ooops very late reply, I think a ValueError makes sense, and amend the documentation.
The other solution would be to make those cURL constants an Enum and enforce that return type but that sounds maybe like overengineering, especially if the enum values might get changed.

But possibly the one where a value of a different type is returned should use the more standard type error message, instead of the "must return X or Y"?